### ISSUE TITLE
RFC: 3 step flow for forgotten password ?

### ISSUE THREAD

**Issue by Ownmarc** [aws-amplify/amplify-js#4876]
## Summary

Would it be possible to validate the code of the forgotten password before having to send the new password ? Maybe its on Cognito's end ?

## Motivation

I didn't find anyway of doing that with amplify. I would not even show the fields to set a new password if I can't validate the user's code for the forgotten password.

Actually there is only 2 function to request a new password : 
- forgotPassword(username) that does its job and shouldn't be touched.
- forgotPasswordSubmit(username, code, newPassword) that I suggest we could split in 2 steps (validate the code first and then allow for password reset instead of sending everything at the same time)

## Basic Example

Being able to make a "forgotten password" flow in 3 steps : 
 - Request code for forgotten password (send username)
- Validate code (send username + code, get back a token or something)
- Set the new password (send the token received at step 2 + username + new password )

## Drawbacks

No, should have its own function along with forgotPasswordSubmit() and forgotPassword()

## Related Issues

Here is all I found that is related : 

https://github.com/amazon-archives/amazon-cognito-identity-js/issues/466

## References

None

**Comment by tnghia944** [comment]
.

**Comment by mlabieniec** [comment]
Have you tried using Cognito Triggers? Specifically the `CustomMessage_ForgotPassword` trigger. This should allow you to retrieve the code/validate and trigger the UI update/s based on the response.

**Comment by Ownmarc** [comment]
@mlabieniec I am not sure I understand how this allow us to validate the code the user is giving us. `CustomMessage_ForgotPassword` is used to template the email/message you will send to the user for forgotten password. You do not get access to the validation code there and even if you did, I do not see how one could use this trigger to do what was described in this issue. Would you mind elaborating ?

**Comment by nick-zh** [comment]
I agree with @Ownmarc, i tried several triggers, i was unable to custom verify the code or get access to the code itself. This would be a very nice feature indeed.
I see two possible approaches:
- be able to verify the code in an individual function (as suggested)
- have a lambda trigger that allows you to customize the code that is being sent, so you have control over it

**Comment by ryan-hellowynd** [comment]
facing the same problem, is there any update on this issue?

**Comment by Zecento** [comment]
any updates?

**Comment by tr3v3r** [comment]
As an hack solution it's possible to do something like.
1)
```javascript
const tempPassword = generateTempPassword()

await Auth.forgotPasswordSubmit(username, code, tempPassword)
// save temp password or pass to next screen
```

```javascript
// on next screen / step just get save temp password and ask user to add new one

await Auth. changePassword(username, tempPassword, newPassword)
```

**Comment by yunchanpaik** [comment]
> As an hack solution it's possible to do something like.
> 
> 
> ```js
> const tempPassword = generateTempPassword()
> 
> await Auth.forgotPasswordSubmit(username, code, tempPassword)
> // save temp password or pass to next screen
> ```
> 
> 
> ```js
> // on next screen / step just get saved temp password and ask user to add new one
> 
> await Auth.changePassword(username, tempPassword, newPassword)
> ```

This method doesn't work anymore since `Auth.changePassword` requires a cognito user instance instead of an username.

Would really love to get an official support from the AWS team on this issue. It's such a frustrating UX to enter the password two times only to find out the code had expired.

**Comment by dawid-bytys** [comment]
any update?

**Comment by dragosheroiu95** [comment]
In our case we have some rules defined for the password (numbers, special chars, uppercase and lowercase letters) and I was able to achieve this by intentionally setting a failing password.
`    try {
      await confirmResetPassword({
        username: email,
        confirmationCode: code,
        newPassword: "thiisafailingpassword", // a passworrd which will always fail since we have restrictions mentioned above
      });
    } catch (error) {
      if (error.name === "InvalidPasswordException") {
        // This means your code is a valid one
      } else {
        // Something is wrong with the code or any other error
        setErrorMessage(error.message);
      }
    }`

**Comment by oemer-aran** [comment]
Can we not just get a new function? Something as simple as `isCodeValid` or `isResetPasswordCodeValid`:
```ts
import { isCodeValid } from "amplify/auth";

await isCodeValid(username, code)
```

**Comment by gaffarmalik** [comment]
Its fascinating that this is still an issue.

I'm trying to see if I can bake the username and code into a JWT link( URL), then allow the frontend to send both (code and new password to me). 
But an issue here would be, how do i validate that the code is not outdated/invalid, incase the user clicks the JWT link in the future.

I'm thinking of using lambda to store code + username + date of creation, so I can let the user know the whether the link is outdated or invalid.

But I'm also trying to avoid making db calls in the aws lamda that sends the token email and the other unnecessary complexities that could accompany this.

**Comment by HuiSF** [comment]
Hi @gaffarmalik thanks for following up, to confirm you are also looking for splitting the OP mentioned `forgotPasswordSubmit` into two separate steps: 1) validating the code 2) resetting the password.

If so, could you provide more details on the reasons why would like this separation?
